File: src/main/java/com/machfour/macros/core/MacrosConfig.java
package com.machfour.macros.core;

import com.machfour.macros.storage.MacrosDataSource;
import com.machfour.macros.storage.MacrosDatabase;

import java.io.BufferedReader;
import java.io.PrintStream;

public interface MacrosConfig {
    // All of these methods return not null

    PrintStream outStream();

    PrintStream errStream();

    BufferedReader inputReader();

    String getInitSqlName();

    String getTrigSqlName();

    String getDataSqlName();

    String getDbName();

    String getFoodCsvPath();

    String getServingCsvPath();

    String getRecipeCsvPath();

    String getIngredientsCsvPath();

    String getDefaultCsvOutputDir();

    String getDbLocation();

    void setDbLocation(String dbLocation);

    String getProgramName();

    // for not-usual operations on the database
    MacrosDatabase getDatabaseInstance();
    MacrosDataSource getDataSourceInstance();
}

File: src/main/kotlin/com/machfour/macros/core/MacrosConfig.kt
package com.machfour.macros.core

import com.machfour.macros.sql.SqlConfig
import com.machfour.macros.sql.SqlDatabase
import com.machfour.macros.sql.SqlDatabaseImpl

interface MacrosConfig {
    val dbName: String
    val foodCsvPath: String
    val servingCsvPath: String
    val recipeCsvPath: String
    val ingredientsCsvPath: String
    var dbLocation: String
    val programName: String

    val sqlConfig: SqlConfig

    val database: SqlDatabase
    // for unusual low-level operations on the database
    val databaseImpl: SqlDatabaseImpl

}